### ISSUE TITLE
[PROPOSAL]: Numeric id-based lineage implementation

### ISSUE THREAD

**Issue by pirz** [microsoft/hyperspace#220]
## Describe the problem
This design proposal is for fixing issue #200

Lineage column in a Hyperspace index tracks the origin of each index entry. This is required to support "delete" during index refresh and HybridScan. Currently, this origin is captured through storing full path to the source data file for each index record.

The data type for the lineage column is `String` which could cause a few issues, when dealing with indexes built on large datasets with many source data files:
* **Storage bloat in index files**: Lineage is a sperate column, added to the index schema. Storing full source data file path for each and every indx record as a String value could increase total size of index files significantly. Although Parquet's built-in compression alleviates this issue, `String` lineage values could still cause storage bloat. As an example, for a set of indexes, created on 1TB of data generated using TPCH dbgen tool, in which each table has up to 100K source data file chunks, the overall storage size for indexes with lineage turned to be more than 23% larger than same indexes with no lineage.

* **High memory footprint when reading the index data**: This is a side effect of storage bloat issue. With larger index files (due to storing lineage as String values), any consumer code s.a. Parquet FileReader ends up experiencing higher memory pressure when scanning and loading index files' content.

* **High comparison cost due to String comparison**: Currently, lineage column values are used to identify deleted index records during refresh in the incremental mode or HybridScan. This requires value check and comparison for each and every record which is costly for the String data type. When scanning and modifying large index files, this can result in considerable slow down in refresh or query time. 

We propose changing the format and data type for the lineage column to reduce the impact of above storage and performance related issues, when working with large indexes. 

### Requirement 

The change should not affect the Actions and functions, using lineage. This means `refreshIndex` in the incremental mode and HybridScan code to support delete should work as is. 

### Solution Overview

We propose to change the data type for the lineage column from String to an integral data type. For this purpose, we assign a unique numeric identifier, as `id`, to each given source data file and store this id, as the lineage value, for index entries coming from that file. This id has below properties:
* For a given index, it is unique per source data file. This means, in an index each and every source data file has a unique id value that remains unchanged throughout the life-time of the source data file in the context of that index. If a given source data file is referred to in the source content of multiple indexes, it could have a different id in each of those indexes.
* If the user deletes a given source data file, its id is not recycled/reused. This means `id` of a deleted file wont be assigned to another source data file, in the context of a given index. 
* The id is for internal-use only. Hyperspace tracks the id values for each index and generates them automatically.

#### Lineage Data Type

We change the data type for the lineage column from `StringType` to the integral type of `LongType`.

According to the [Spark documentation](https://spark.apache.org/docs/latest/sql-ref-datatypes.html): 
>> `LongType` represents 8-byte signed integer numbers, with the range of numbers from -9223372036854775808 to 9223372036854775807.

As Hyperspace generates the source data files' `id` values automatically and uses them internally, we can use both negative and positive values. However, for readability, we start by using the positive range of numbers (starting at 0). Numbers from the positive range should be (more than) enough to cover all source data files that could appear in the source content of an index throughout its life-time. Sticking to positive values, we ideally prefer an 8-byte unsigned integral value, `uint64`. However, Spark does not have a 64-bit unsigned integer data type and `LongType` is the closet we can get to it. We emphasize that nothing prevents us from using the numbers from the negative range. The choice of starting from 0 and using postive numbers is only for readability and simple debuggability.

#### Id Generation

Hyperspace generates the `id` values for source data files automatically.

Once an index is created, Hyperspace starts assigning `id` values to the source data files listed in the source content. The id values starts from 0 and is monotonically increasing as new source data files are identified for the index.

A new source data file is detected during (1) index creation, (2) full refresh or (3) incremental refresh. During each of these actions, Hyperspace generates a new id for each new source data file and assign it to that file. This id is saved in the `FileInfo` for the new file and remains immutable (see below). 

#### Index Metadata Changes

Hyperspace uses index metadata to track id generation process by persisting last used id value for the source data files of a given index. For this purpose, we propose below index metadata changes in the `IndexLogEntry` class:

* A new field, named `maxId` is added under `source.plan.properties`. Note that this tracks the `id` generation among all `relations` in an index by storing maximum used id for source data files so far. As a results, `SparkPlan` class will look like as:

```
object SparkPlan {
  case class Properties(
      relations: Seq[Relation],
      rawPlan: String,
      sql: String,
      fingerprint: LogicalPlanFingerprint,
      maxId: Long   // New field to track max source data file id, so far
      )
}
```
* The `FileInfo` class is modified to store the generated `id` associated to each source data file.

```
case class FileInfo(
    name: String,
    size: Long,
    modifiedTime: Long,
    id: Long    // New field added to store auto-generated id for this file
    )
```

* The data type for the lineage column, named as `_data_file_name` is changed from `StringType` to `LongType`. In the metadata, this will show up as part of `schemaString` in `CoveringIndex.Properties`. 

#### Correctness
All existing tests in the Hyperspace test suite for `CreateIndexTests`, `RefreshIndexTests` and `E2EHyperspaceRulesTests` should pass as is.

#### Storage Bloat
We will compare storage bloat for recommended indexes on data generated with TPCH dbgen tool on SF=1000GB. We compare the storage bloat for the case lineage is captured as `Long` vs. the case it is captured as `String`. The original tables' data are stored in two layouts: "up to 200 source data files" and "up to 100K source data files".  

#### Incremental Refresh Performance
Using same data as the Storage Bloat experiment above, we measure the performance of incremental refresh action when we delete 10%, 20% or 50% of original data files for the `LineItem` table (from the TPCH schema). We use "100K Source data files" in this experiment. This measurement will be done on both lineage stored as `String` and stored as `Long` to compare them to each other.

#### Query Performance
We use HybridScan, with "delete" enabled, for this experiment. We compare the performance of selected queries, derived from TPCH specification, on the data used in the Incremental Refresh Performance experiment above and see how the query performance changes when HybridScan enforces "delete" once 10%, 20% or 50% of the original data files are deleted from the `LineItem` table. This comparison happens between lineage stored as `String` and stored as `Long`.

## Describe alternatives you've considered

NA

## Additional context

NA

**Comment by sezruby** [comment]
What about the column name? Should we change the name or no? (`_data_file_id`?)

**Comment by pirz** [comment]
Implemented in #234
